pyboy/core/cpu.py

```diff
@@ -168,35 +89,107 @@ def load_state(self, f, state_version):
         self.interrupt_master_enable = f.read()
         self.halted = f.read()
         self.stopped = f.read()
-        logger.debug(
-            f"State loaded: A:{self.A:02x}, F:{self.F:02x}, B:{self.B:02x}, C:{self.C:02x}, D:{self.D:02x}, E:{self.E:02x}, HL:{self.HL:02x}, SP:{self.SP:02x}, PC:{self.PC:02x}, IME:{self.interrupt_master_enable}, halted:{self.halted}, stopped:{self.stopped}"
+        if state_version >= 5:
+            # Interrupt register moved from RAM to CPU
+            self.interrupts_enabled_register = f.read()
+        logger.debug("State loaded: " + self.dump_state())
+
+    def dump_state(self):
+        return (
+            f"A:{self.A:02x}, F:{self.F:02x}, B:{self.B:02x}, C:{self.C:02x}, D:{self.D:02x}, E:{self.E:02x}, "
+            f"HL:{self.HL:02x}, SP:{self.SP:02x}, PC:{self.PC:02x}, "
+            f"Opcode:{self.mb.getitem(self.PC):02x}, "
+            f"Opcode+1:{self.mb.getitem((self.PC+1) & 0xFFFF):02x}, "
+            f"Opcode+2:{self.mb.getitem((self.PC+2) & 0xFFFF):02x}, "
+            f"IME:{self.interrupt_master_enable}, halted:{self.halted}, "
+            f"interrupt_queued:{self.interrupt_queued} "
+            f"stopped:{self.stopped}"
         )
 
+    def set_interruptflag(self, flag):
+        self.interrupts_flag_register |= flag
+
+    def tick(self):
+        if self.check_interrupts():
+            self.halted = False
+            # TODO: We return with the cycles it took to handle the interrupt
+            return 0
+
+        if self.halted and self.interrupt_queued:
+            # GBCPUman.pdf page 20
+            # WARNING: The instruction immediately following the HALT instruction is "skipped" when interrupts are
+            # disabled (DI) on the GB,GBP, and SGB.
+            self.halted = False
+            self.PC += 1
+            self.PC &= 0xFFFF
+        elif self.halted:
+            return 4 # TODO: Number of cycles for a HALT in effect?
+
+        old_pc = self.PC
+        cycles = self.fetch_and_execute(self.PC)
+        if not self.halted and old_pc == self.PC and not self.is_stuck:
+            logger.error("CPU is stuck: " + self.dump_state())
+            self.is_stuck = True
+        self.interrupt_queued = False
+        return cycles
+
+    def check_interrupts(self):
+        if self.interrupt_queued:
+            # Interrupt already queued. This happens only when using a debugger.
+            return False
+
+        if (self.interrupts_flag_register & 0b11111) & (self.interrupts_enabled_register & 0b11111):
```

Comment:
```suggestion
        if (self.interrupts_flag_register & self.interrupts_enabled_register & 0b11111):
```

Reply:
I can see what you want to do, but I'm worried someone (me) won't be able to immediately recognize the semantics of this, and might need to be explained with a comment. And the compiler will optimize it anyway.

Reply:
Yeah no worries; I'll drop this one